rfcs/0004-range-queries.md

```diff
@@ -0,0 +1,219 @@
+This includes the active WAL, memtables waiting to be flushed, and L0/SR tables 
+which may need to be loaded from object storage. Internally, we will rely on 
+iterators which sort and merge these tables.  The latest records will
+have precedence following the same rules that a normal `get` query follows.
+
+The existing `MergeIterator` used for compaction provides much of the necessary
+foundation to unify multiple iterators. Additionally, `MemTableIterator` and 
+`SstIterator` provide ready-to-use capabilities to scan tables which may keys
+within the range. 
+
+**Query Isolation**: When using the `Uncommitted` isolation, unflushed
+records recorded in WALs will be used as a source for the key range. Otherwise,
+only committed records in memtables, as well as L0 and SR tables will be used.
+
+Range queries must access mutable table state, such as the current WAL. It is 
+therefore possible for the query results to include some records which 
+were written after the  query began scanning its range of records. For example,
+a query may include  only a subset of records that were inserted in the same 
+`WriteBatch`. In  other words, range queries will not have snapshot isolation 
+initially. 
+
+[RFC-????](https://github.com/slatedb/slatedb/pull/260) proposes transactional
+semantics for SlateDB. This relies on a monotonic sequence number which is tagged
+on each key update. Range queries could leverage this sequence number to improve
+isolation by filtering any updates with sequence numbers larger than the latest
+at the time the query was executed. We leave this as a gap for future work to address.
+
+### Checkpoint References
+
+We intend to rely on state checkpoints introduced in [RFC-????](https://github.com/slatedb/slatedb/pull/278) in order to prevent the cleanup
+of durable state referenced by the query. When a range scan is created, it will create
+an internal reference to the current checkpoint. This reference will prevent cleanup of
+the checkpoint state until the scan is complete. 
+
+Checkpoints created by `Readers` have an expiry time, which is only refreshed as long
+as it remains the current checkpoint. If a `DbIterator` instance remains active long 
+enough for its referenced checkpoint to expire, the `Reader` will proactively invalidate 
+it in  order to allow for safe cleanup. Invalidated iterators will return 
+`SlateDbError::InvalidatedIterator` on any subsequent call to `DbIterator::next`
+or `DbIterator::seek`.
+
+In general, we make the initial simplifying assumption that scans will not be frequent
+enough or last long enough for checkpoint accumulation to be a problem. In the future,
+we could either limit the number of checkpoint references that are available for
+scanning or we could limit the time that a `DbIterator` will remain valid.
+
+### Memory Usage
+
+In addition to the checkpoint, a range scan may hold references to records in unflushed 
```

Comment:
Interesting... this seems loosely related to #277 ... IIUC, you're saying that you'd point to a subset of the WAL/memtable rather than the entire thing ... I think this might be similar to option 3 in #277 that I mentioned ... 🤔 Just thinking out loud. Not any real implication to this RFC...

Reply:
We could count this as a _potential_ optimization in the RFC and sort of start with the naive approach. It seemed likely that most range queries would target a relatively small number of keys, so I thought it might make sense to just grab the matches from memory instead of holding everything.